packages/reactor/docs/planning/Operations/index.md

```diff
@@ -0,0 +1,223 @@
+
+  /** The context of the action */
+  context: ActionContext;
+
+  /** The attachments used in the action */
+  attachments: AttachmentRef[];
+};
+```
+
+#### Action Signing
+
+Actions include cryptographic signatures on the `ActionContext` to verify that they originated from a specific user.
+
+- Cryptographic signatures are verified using the Web Crypto API.
+- Signatures need: a public key, a payload, and a signature.
+- The signature payload is generated from the `SignedPaylodParameters` and the `Action` using a stable, JSON-encoding.
+  - This allows us to dedupe information.
+  - Allows for wrapping signatures in other signatures.
+  - Allows for different payloads, depending on the use case. [Upgrade Actions](./upgrades.md), for instance, may have a different payload than a regular Action.
+
+```tsx
+
+type PublicKey {
+  /**
+   * The name of the algorithm used to sign the action.
+   * 
+   * See: https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/verify#algorithm
+   */
+  algorithm: string;
+
+  /**
+   * The format of the public key.
+   * 
+   * https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/exportKey#format
+   */
+  format: string;
+
+  /**
+   * The exported public key used to sign the action, in Base64 format.
+   */
+  data: string;
+}
+
+type Signer = {
+  /**
+   * The public key used to sign the action.
+   */
+  publicKey: PublicKey;  // public key
+
+  /**
+   * The name of the signer, used to identify a signing application.
+   * 
+   * Eg: "Connect", "Powerhouse", etc
+   */
+  app?: string;
+  
+  /**
+   * EVM address of the signer, used to identify a user.
+   */
+  address?: string;
+
+  /**
+   * The CAIP-2 network id of the signer, used to identify a user.
+   */
+  networkId?: string;
+
+  /**
+   * The CAIP-10 chain id of the signer, used to identify a user.
+   */
+  chainId?: number;
+
+  // todo: other claims?
+};
+
+type SignedPayloadParameters = {
+  /**
+   * Additional signatures to append to the payload.
+   */
+  signatures: string[];
+};
+
+type Signature = {
+  /**
+   * The signer of the action.
+   */
+  signer: Signer;
+
+  /**
+   * Parameters for the signed payload.
+   */
+  payload: SignedPayloadParameters;
+
+  /**
+   * Base64 encoded signature of the payload.
+   */
+  signature: string;
+};
+
+type ActionContext = {
+  /**
+   * The signature of the action.
+   */
+  signature: Signature;
+};
+```
+
+### Operations
+
+An `Operation` represents the result of an `Action` being executed. It is a "fact", not an "intent".
+
+Unlike a typical ES approach, where Actions are generally discarded or stored in an outbox pattern, each `Action` is stored with the `Operation` that is produced from it. This allows every client to re-execute each `Action` and verify the resulting `Operation`.
+
+The lack of `previousState` is simply an optimization. As the previous state is always known, it is not necessary to store it (but perhaps could be necessary for debugging).
```

Comment:
I'd also remove `resultingState` from the `Operation` type as it can be very expensive.

Reply:
If we put the state only in the document view, then because it's only a projection of the operations store, it does not have the state information unless it replays the operations to get it. It would be tempting just to pass the state through the in-memory event bus to the view, but this doesn't really help out as there are always cases in which we replay operations. This means we'd be double applying operations: once in the job executor and once in the document view.

We keep saying "Write Lock" but in reality this will simply use optimistic locking, so it should not be an issue.

Reply:
1. Optimistic locking - I'll add a section that details the alternatives and then we can bring it up with Prometheus.

2. Dispatched events - let me add a section to the spec on that where we can discuss in more detail.

3. CPU bottleneck - good to know. As the job executor is built to be the thing that scales out horizontally, putting the load there is ideal. Having good benchmarks we can compare commit-to-commit will be essential.

Reply:
Links:

- For locking considerations, see [IOperationsStore].
- For dispatched actions, see the [System Stream](https://github.com/powerhouse-inc/powerhouse/pull/1573/files#diff-92412a25905a1627010bc145b8235dece1b9482e8dcf5ad3a5554baa78cd5c80R223) and [Action Dispatches](https://github.com/powerhouse-inc/powerhouse/pull/1573/files#diff-92412a25905a1627010bc145b8235dece1b9482e8dcf5ad3a5554baa78cd5c80R244) sections of the Operations spec.